README.md
PgBouncer
=========

This is a fork of the pgbouncer that adds support for multiprocessing. It is based on the 1.12.0 version of pgbouncer.

When enabled pgbouncer will fork itself into a manager process
and multiple worker processes. The manager then waits for new connections on
the specified `listen_addr` and then sends the connected client socket to one of
the workers to handle all the requests.

Building
---------

SMP is enabled by default, but can be disabled by specifying:

    $ ./configure --without-smp ...

PgBouncer depends on few things to get compiled:

* [GNU Make] 3.81+
* [Libevent] 2.0+
* [pkg-config]
* [OpenSSL] 1.0.1+ for TLS support
* (optional) [c-ares] as alternative to Libevent's evdns
* (optional) PAM libraries

[GNU Make]: https://www.gnu.org/software/make/
[Libevent]: http://libevent.org/
[pkg-config]: https://www.freedesktop.org/wiki/Software/pkg-config/
[OpenSSL]: https://www.openssl.org/
[c-ares]: http://c-ares.haxx.se/

When dependencies are installed just run:

    $ ./configure --prefix=/usr/local
    $ make
    $ make install

If you are building from Git, or are building for Windows, please see
separate build instructions below.

DNS lookup support
------------------

PgBouncer does host name lookups at connect time instead of just once
at configuration load time.  This requires an asynchronous DNS
implementation.  The following table shows supported backends and
their probing order:

| backend                    | parallel | EDNS0 (1) | /etc/hosts | SOA lookup (2) | note                                  |
|----------------------------|----------|-----------|------------|----------------|---------------------------------------|
| c-ares                     | yes      | yes       | yes        | yes            | IPv6+CNAME buggy in <=1.10            |
| udns                       | yes      | yes       | no         | yes            | IPv4 only                             |
| evdns, libevent 2.x        | yes      | no        | yes        | no             | does not check /etc/hosts updates     |
| getaddrinfo_a, glibc 2.9+  | yes      | yes (3)   | yes        | no             | N/A on non-glibc                      |
| getaddrinfo, libc          | no       | yes (3)   | yes        | no             | N/A on Windows, requires pthreads     |

1. EDNS0 is required to have more than 8 addresses behind one host name.
2. SOA lookup is needed to re-check host names on zone serial change.
3. To enable EDNS0, add `options edns0` to `/etc/resolv.conf`.

c-ares is the most fully-featured implementation and is recommended
for most uses and binary packaging (if a sufficiently new version is
available).  Libevent's built-in evdns is also suitable for many uses,
with the listed restrictions.  The other backends are mostly legacy
options at this point and don't receive much testing anymore.

By default, c-ares is used if it can be found.  Its use can be forced
with `configure --with-cares` or disabled with `--without-cares`.  If
c-ares is not used (not found or disabled), then specify `--with-udns`
to pick udns, else Libevent is used.  Specify `--disable-evdns` to
disable the use of Libevent's evdns and fall back to a libc-based
implementation.

PAM authentication
------------------

To enable PAM authentication, `./configure` has a flag `--with-pam`
(default value is no).  When compiled with PAM support, a new global
authentication type `pam` is available to validate users through PAM.

Building from Git
-----------------

Building PgBouncer from Git requires that you fetch the libusual
submodule and generate the header and configuration files before
you can run `configure`:

	$ git clone https://github.com/pgbouncer/pgbouncer.git
	$ cd pgbouncer
	$ git submodule init
	$ git submodule update
	$ ./autogen.sh
	$ ./configure ...
	$ make
	$ make install

Additional packages required: autoconf, automake, libtool, pandoc

Building on Windows
-------------------

The only supported build environment on Windows is MinGW.  Cygwin and
Visual $ANYTHING are not supported.

To build on MinGW, do the usual:

	$ ./configure ...
	$ make

If cross-compiling from Unix:

	$ ./configure --host=i586-mingw32msvc ...

Running on Windows
------------------

Running from the command line goes as usual, except that the `-d` (daemonize),
`-R` (reboot), and `-u` (switch user) switches will not work.

To run PgBouncer as a Windows service, you need to configure the
`service_name` parameter to set name for service.  Then:

	$ pgbouncer -regservice config.ini

To uninstall service:

	$ pgbouncer -unregservice config.ini

To use the Windows event log, set `syslog = 1` in the configuration file.
But before that you need to register `pgbevent.dll`:

	$ regsvr32 pgbevent.dll

To unregister it, do:

	$ regsvr32 /u pgbevent.dll


debian/changelog
pgbouncer (1.8.1-2) unstable; urgency=low

  * v1.8.1-2

 -- Peter Eisentraut <peter@eisentraut.org>  Wed, 20 Dec 2017 00:00:00 +0000
pgbouncer (1.8.1-1) unstable; urgency=low

  * v1.8.1

 -- Peter Eisentraut <peter@eisentraut.org>  Wed, 20 Dec 2017 00:00:00 +0000

pgbouncer (1.8-1) unstable; urgency=low

  * v1.8

 -- Peter Eisentraut <peter@eisentraut.org>  Tue, 19 Dec 2017 00:00:00 +0000

pgbouncer (1.7.2-1) unstable; urgency=low

  * v1.7.2

 -- Marko Kreen <markokr@gmail.com>  Fri, 26 Feb 2016 10:29:47 +0200

pgbouncer (1.7.1-1) unstable; urgency=low

  * v1.7.1

 -- Marko Kreen <markokr@gmail.com>  Thu, 18 Feb 2016 18:34:55 +0200

pgbouncer (1.7-1) unstable; urgency=low

  * v1.7

 -- Marko Kreen <markokr@gmail.com>  Fri, 18 Dec 2015 19:57:16 +0200

pgbouncer (1.7~rc1-1) unstable; urgency=low

  * 1.7rc1

 -- Marko Kreen <markokr@gmail.com>  Mon, 02 Nov 2015 18:11:58 +0200

pgbouncer (1.7~dev-1) unstable; urgency=low

  * v1.7dev

 -- Marko Kreen <markokr@gmail.com>  Mon, 03 Aug 2015 23:47:46 +0300

pgbouncer (1.6-1) unstable; urgency=low

  * v1.6

 -- Petr Jelinek <pjmodos@pjmodos.net>  Sat, 01 Aug 2015 11:53:28 +0200

pgbouncer (1.6rc1-1) unstable; urgency=low

  * v1.6rc1

 -- Marko Kreen <markokr@gmail.com>  Tue, 30 Dec 2014 16:19:51 +0200

pgbouncer (1.6~dev-1) unstable; urgency=low

  * v1.6dev

 -- Marko Kreen <markokr@gmail.com>  Tue, 08 Oct 2013 00:37:48 +0300

pgbouncer (1.5.4-1) unstable; urgency=low

  * v1.5.4

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 28 Nov 2012 12:44:39 +0200

pgbouncer (1.5.3-1) unstable; urgency=low

  * v1.5.3

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 12 Sep 2012 13:36:51 +0300

pgbouncer (1.5.2-1) unstable; urgency=low

  * v1.5.2

 -- Marko Kreen <marko.kreen@skype.net>  Tue, 29 May 2012 11:02:48 +0300

pgbouncer (1.5.1-1) unstable; urgency=low

  * v1.5.1

 -- Marko Kreen <marko.kreen@skype.net>  Tue, 17 Apr 2012 16:04:13 +0300

pgbouncer (1.5.1rc1-1) unstable; urgency=low

  * v1.5.1rc1

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 14 Mar 2012 14:31:54 +0200

pgbouncer (1.5-1) unstable; urgency=low

  * v1.5

 -- Marko Kreen <marko.kreen@skype.net>  Thu, 05 Jan 2012 14:59:43 +0200

pgbouncer (1.5rc1-1) unstable; urgency=low

  * v1.5rc1

 -- Marko Kreen <marko.kreen@skype.net>  Tue, 13 Dec 2011 00:28:43 +0200

pgbouncer (1.4.3dev1-1) unstable; urgency=low

  * v1.4.3dev1

 -- Marko Kreen <marko.kreen@skype.net>  Thu, 01 Sep 2011 11:49:12 +0300

pgbouncer (1.4.2rc1-1) unstable; urgency=low

  * v1.4.2rc1

 -- Marko Kreen <marko.kreen@skype.net>  Thu, 02 Jun 2011 18:36:27 +0300

pgbouncer (1.4.1-1) unstable; urgency=low

  * v1.4.1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 01 Apr 2011 21:51:54 +0300

pgbouncer (1.4.1rc5-1) unstable; urgency=low

  * v1.4.1rc5

 -- Marko Kreen <marko.kreen@skype.net>  Tue, 29 Mar 2011 17:00:55 +0300

pgbouncer (1.4.1rc3-1) unstable; urgency=low

  * v1.4.1rc4

 -- Marko Kreen <marko.kreen@skype.net>  Sat, 26 Mar 2011 00:40:50 +0200

pgbouncer (1.4.1rc3-1) unstable; urgency=low

  * v1.4.1rc3

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 23 Mar 2011 18:06:01 +0200

pgbouncer (1.4.1rc2-1) unstable; urgency=low

  * v1.4.1rc2

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 21 Mar 2011 23:24:00 +0200

pgbouncer (1.4.1rc1-1) unstable; urgency=low

  * v1.4.1rc1

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 28 Feb 2011 23:11:27 +0200

pgbouncer (1.4-1) unstable; urgency=low

  * v1.4

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 08 Dec 2010 17:30:06 +0200

pgbouncer (1.4-0rc4) unstable; urgency=low

  * v1.4rc4

 -- Marko Kreen <marko.kreen@skype.net>  Tue, 30 Nov 2010 11:54:55 +0200

pgbouncer (1.4-0rc3) unstable; urgency=low

  * v1.4rc3

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 29 Nov 2010 13:44:46 +0200

pgbouncer (1.4-0rc2.1) unstable; urgency=low

  * v1.4rc2.1

 -- Marko Kreen <marko.kreen@skype.net>  Thu, 18 Nov 2010 10:26:25 +0200

pgbouncer (1.4-0rc2) unstable; urgency=low

  * v1.4rc2

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 17 Nov 2010 15:11:13 +0200

pgbouncer (1.4-0rc1.1) unstable; urgency=low

  * v1.4rc1.1

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 17 Nov 2010 14:10:25 +0200

pgbouncer (1.4-0rc1) unstable; urgency=low

  * v1.4rc1

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 08 Nov 2010 11:02:57 +0200

pgbouncer (1.4-0rc0) unstable; urgency=low

  * v1.4rc0

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 03 Nov 2010 15:14:10 +0200

pgbouncer (1.4-0alpha1) unstable; urgency=low

  * v1.4 alpha 1

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 20 Sep 2010 13:37:15 -0700

pgbouncer (1.3.3-0rc1) unstable; urgency=low

  * v1.3.3rc1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 23 Apr 2010 17:33:07 +0300

pgbouncer (1.3.2-1) unstable; urgency=low

  * v1.3.2

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 15 Mar 2010 16:09:16 +0200

pgbouncer (1.3.2rc1-1) unstable; urgency=low

  * v1.3.2rc1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 04 Dec 2009 13:18:26 +0200

pgbouncer (1.3.1-1) unstable; urgency=low

  * v1.3.1

 -- Marko Kreen <marko.kreen@skype.net>  Mon, 06 Jul 2009 16:12:53 +0300

pgbouncer (1.3.1rc1-1) unstable; urgency=low

  * v1.3.1rc1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 26 Jun 2009 14:25:42 +0300

pgbouncer (1.3-1) unstable; urgency=low

  * v1.3

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 18 Feb 2009 14:11:41 +0200

pgbouncer (1.3rc1-0) unstable; urgency=low

  * v1.3rc1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 16 jan 2009 15:28:49 +0200

pgbouncer (1.2.3-1) unstable; urgency=low

  * v.1.2.3

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 08 Aug 2008 14:35:27 +0300

pgbouncer (1.2.2-1) unstable; urgency=low

  * v1.2.2

 -- Marko Kreen <marko.kreen@skype.net>  Wed, 06 Aug 2008 09:44:15 +0300

pgbouncer (1.2.1-1) unstable; urgency=low

  * v1.2.1

 -- Marko Kreen <marko.kreen@skype.net>  Fri, 01 Aug 2008 13:00:30 +0300

pgbouncer (1.2-1) unstable; urgency=low

  * v1.2


src/util.c
/*
 * PgBouncer - Lightweight connection pooler for PostgreSQL.
 *
 * Copyright (c) 2007-2009  Marko Kreen, Skype Technologies OÜ
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Random small utility functions
 */

#include "bouncer.h"

#include <usual/crypto/md5.h>
#include <usual/crypto/csrandom.h>

int log_socket_prefix(enum LogLevel lev, void *ctx, char *dst, unsigned int dstlen)
{
	const struct PgSocket *sock = ctx;
	const char *user, *db, *host;
	char host6[PGADDR_BUF];
	int port;
	char stype;

	/* no prefix */
	if (!sock)
		return 0;

	/* format prefix */
	stype = is_server_socket(sock) ? 'S' : 'C';
	port = pga_port(&sock->remote_addr);
	db = sock->pool ? sock->pool->db->name : "(nodb)";
	user = sock->auth_user ? sock->auth_user->name : "(nouser)";
	if (pga_is_unix(&sock->remote_addr)) {
		unsigned long pid = sock->remote_addr.scred.pid;
		if (pid) {
			snprintf(host6, sizeof(host6), "unix(%lu)", pid);
			host = host6;
		} else {
			host = "unix";
		}
	} else {
		host = pga_ntop(&sock->remote_addr, host6, sizeof(host6));
	}

	if (pga_family(&sock->remote_addr) == AF_INET6) {
		return snprintf(dst, dstlen, "%c-%p: %s/%s@[%s]:%d ",
			stype, sock, db, user, host, port);
	} else {
		return snprintf(dst, dstlen, "%c-%p: %s/%s@%s:%d ",
			stype, sock, db, user, host, port);
	}
}

const char *bin2hex(const uint8_t *src, unsigned srclen, char *dst, unsigned dstlen)
{
	unsigned int i, j;
	static const char hextbl [] = "0123456789abcdef";
	if (!dstlen)
		return "";
	if (srclen*2+1 > dstlen)
		srclen = (dstlen - 1) / 2;
	for (i = j = 0; i < srclen; i++) {
		dst[j++] = hextbl[src[i] >> 4];
		dst[j++] = hextbl[src[i] & 15];
	}
	dst[j] = 0;
	return dst;
}

/*
 * PostgreSQL MD5 hashing.
 */

static void hash2hex(const uint8_t *hash, char *dst)
{
	bin2hex(hash, MD5_DIGEST_LENGTH, dst, 16*2+1);
}

void pg_md5_encrypt(const char *part1,
		    const char *part2, size_t part2len,
		    char *dest)
{
	struct md5_ctx ctx;
	uint8_t hash[MD5_DIGEST_LENGTH];

	md5_reset(&ctx);
	md5_update(&ctx, part1, strlen(part1));
	md5_update(&ctx, part2, part2len);
	md5_final(&ctx, hash);

	memcpy(dest, "md5", 3);
	hash2hex(hash, dest + 3);
}

/* wrapped for getting random bytes */
void get_random_bytes(uint8_t *dest, int len)
{
	csrandom_bytes(dest, len);
}

/* set needed socket options */
bool tune_socket(int sock, bool is_unix)
{
	int res;
	int val;
	bool ok;

	/*
	 * Generic stuff + nonblock.
	 */
	ok = socket_setup(sock, true);
	if (!ok)
		goto fail;

	/*
	 * Following options are for network sockets
	 */
	if (is_unix)
		return true;

	/*
	 * TCP Keepalive
	 */
	ok = socket_set_keepalive(sock, cf_tcp_keepalive, cf_tcp_keepidle,
				  cf_tcp_keepintvl, cf_tcp_keepcnt);
	if (!ok)
		goto fail;

	/*
	 * set in-kernel socket buffer size
	 */
	if (cf_tcp_socket_buffer) {
		val = cf_tcp_socket_buffer;
		res = setsockopt(sock, SOL_SOCKET, SO_SNDBUF, &val, sizeof(val));
		if (res < 0)
			goto fail;
		val = cf_tcp_socket_buffer;
		res = setsockopt(sock, SOL_SOCKET, SO_RCVBUF, &val, sizeof(val));
		if (res < 0)
			goto fail;
	}

	/*
	 * Turn off kernel buffering, each send() will be one packet.
	 */
	val = 1;
	res = setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &val, sizeof(val));
	if (res < 0)
		goto fail;
	return true;
fail:
	log_warning("tune_socket(%d) failed: %s", sock, strerror(errno));
	return false;
}

/*
 * Find a string in comma-separated list.
 *
 * It does not support space inside tokens.
 */
bool strlist_contains(const char *liststr, const char *str)
{
	int c, len = strlen(str);
	const char *p, *listpos = liststr;

loop:
	/* find string fragment, later check if actual token */
	p = strstr(listpos, str);
	if (p == NULL)
		return false;

	/* move listpos further */
	listpos = p + len;
	/* survive len=0 and avoid unnecessary compare */
	if (*listpos)
		listpos++;

	/* check previous symbol */
	if (p > liststr) {
		c = *(p - 1);
		if (!isspace(c) && c != ',')
			goto loop;
	}

	/* check following symbol */
	c = p[len];
	if (c != 0 && !isspace(c) && c != ',')
		goto loop;

	return true;
}

void fill_remote_addr(PgSocket *sk, int fd, bool is_unix)
{
	PgAddr *dst = &sk->remote_addr;
	socklen_t len = sizeof(PgAddr);
	int err;

	if (is_unix) {
		uid_t uid = 0;
		gid_t gid = 0;
		pid_t pid = 0;
		pga_set(dst, AF_UNIX, cf_listen_port);
		if (getpeercreds(fd, &uid, &gid, &pid) >= 0) {
			log_noise("unix peer uid: %d", (int)uid);
		} else {
			log_warning("unix peer uid failed: %s", strerror(errno));
		}
		dst->scred.uid = uid;
		dst->scred.pid = pid;
	} else {
		err = getpeername(fd, (struct sockaddr *)dst, &len);
		if (err < 0) {
			log_error("fill_remote_addr: getpeername(%d) = %s",
				  fd, strerror(errno));
		}
	}
}

void fill_local_addr(PgSocket *sk, int fd, bool is_unix)
{
	PgAddr *dst = &sk->local_addr;
	socklen_t len = sizeof(PgAddr);
	int err;

	if (is_unix) {
		pga_set(dst, AF_UNIX, cf_listen_port);
		dst->scred.uid = geteuid();
		dst->scred.pid = getpid();
	} else {
		err = getsockname(fd, (struct sockaddr *)dst, &len);
		if (err < 0) {
			log_error("fill_local_addr: getsockname(%d) = %s",
				  fd, strerror(errno));
		}
	}
}

/*
 * Error handling around evtimer_add() is nasty as the code
 * may not be called again.  As there is fixed number of timers
 * in pgbouncer, provider safe_evtimer_add() that stores args of
 * failed calls in static array and retries later.
 */
#define TIMER_BACKUP_SLOTS  10

struct timer_slot {
	struct event *ev;
	struct timeval tv;
};
static struct timer_slot timer_backup_list[TIMER_BACKUP_SLOTS];
static int timer_backup_used = 0;

void safe_evtimer_add(struct event *ev, struct timeval *tv)
{
	int res;
	struct timer_slot *ts;

	res = evtimer_add(ev, tv);
	if (res >= 0)
		return;

	if (timer_backup_used >= TIMER_BACKUP_SLOTS)
		fatal_perror("TIMER_BACKUP_SLOTS full");

	ts = &timer_backup_list[timer_backup_used++];
	ts->ev = ev;
	ts->tv = *tv;
}

void rescue_timers(void)
{
	struct timer_slot *ts;
	while (timer_backup_used) {
		ts = &timer_backup_list[timer_backup_used - 1];
		if (evtimer_add(ts->ev, &ts->tv) < 0)
			break;
		timer_backup_used--;
	}
}


/*
 * PgAddr operations
 */

int pga_port(const PgAddr *a)
{
	if (a->sa.sa_family == AF_INET6) {
		return ntohs(a->sin6.sin6_port);
	} else {
		return ntohs(a->sin.sin_port);
	}
}

/* set family and port */
void pga_set(PgAddr *a, int af, int port)
{
	memset(a, 0, sizeof(*a));
	if (af == AF_INET6) {
		a->sin6.sin6_family = af;
		a->sin6.sin6_port = htons(port);
	} else {
		a->sin.sin_family = af;
		a->sin.sin_port = htons(port);
	}
}

/* copy sockaddr_in/in6 to PgAddr */
void pga_copy(PgAddr *a, const struct sockaddr *sa)
{
	switch (sa->sa_family) {
	case AF_INET:
		memcpy(&a->sin, sa, sizeof(a->sin));
		break;
	case AF_INET6:
		memcpy(&a->sin6, sa, sizeof(a->sin6));
		break;
	case AF_UNIX:
		log_error("pga_copy: AF_UNIX copy not supported");
	}
}

int pga_cmp_addr(const PgAddr *a, const PgAddr *b)
{
    if (pga_family(a) != pga_family(b))
		return pga_family(a) - pga_family(b);

	switch (pga_family(a)) {
	case AF_INET:
		return memcmp(&a->sin.sin_addr, &b->sin.sin_addr, sizeof(a->sin.sin_addr));
		break;
	case AF_INET6:
		return memcmp(&a->sin6.sin6_addr, &b->sin6.sin6_addr, sizeof(a->sin6.sin6_addr));
		break;
	default:
		log_error("pga_cmp_addr: unsupported family");
		return 0;
	}
}

/* convert pgaddr to string */
const char *pga_ntop(const PgAddr *a, char *dst, int dstlen)
{
	const char *res = NULL;
	char buf[PGADDR_BUF];

	memset(buf, 0, sizeof(buf));

	switch (pga_family(a)) {
	case AF_UNIX:
		res = "unix";
		break;
	case AF_INET:
		res = inet_ntop(AF_INET, &a->sin.sin_addr, buf, sizeof(buf));
		break;
	case AF_INET6:
		res = inet_ntop(AF_INET6, &a->sin6.sin6_addr, buf, sizeof(buf));
		break;
	default:
		res = "(bad-af)";
	}
	if (res == NULL)
		res = "(err-ntop)";

	strlcpy(dst, res, dstlen);
	return dst;
}

/* parse address from string */
bool pga_pton(PgAddr *a, const char *s, int port)
{
	int res = 1;
	if (strcmp(s, "unix") == 0) {
		pga_set(a, AF_UNIX, port);
	} else if (strcmp(s, "*") == 0) {
		pga_set(a, AF_INET, port);
		a->sin.sin_addr.s_addr = htonl(INADDR_ANY);
	} else if (strchr(s, ':')) {
		pga_set(a, AF_INET6, port);
		res = inet_pton(AF_INET6, s, &a->sin6.sin6_addr);
	} else {
		pga_set(a, AF_INET, port);
		res = inet_pton(AF_INET, s, &a->sin.sin_addr);
	}
	if (res == 0)
		errno = EINVAL;
	return res > 0;
}

const char *pga_str(const PgAddr *a, char *dst, int dstlen)
{
	char buf[PGADDR_BUF];
	pga_ntop(a, buf, sizeof(buf));
	if (pga_family(a) == AF_INET6) {
		snprintf(dst, dstlen, "[%s]:%d", buf, pga_port(a));
	} else if (pga_family(a) == AF_UNIX && a->scred.pid) {
		snprintf(dst, dstlen, "%s:%d$%lu", buf, pga_port(a), (unsigned long)a->scred.pid);
	} else {
		snprintf(dst, dstlen, "%s:%d", buf, pga_port(a));
	}
	return dst;
}

static const char *cached_hostname(void)
{
	static char cache[256];
	int err;

	if (cache[0] == 0) {
		err = gethostname(cache, sizeof(cache));
		if (err != 0)
			strlcpy(cache, "somehost", sizeof(cache));
	}
	return cache;
}

const char *pga_details(const PgAddr *a, char *dst, int dstlen)
{
	char buf[PGADDR_BUF];
	pga_ntop(a, buf, sizeof(buf));
	if (pga_family(a) == AF_INET6) {
		snprintf(dst, dstlen, "[%s]:%d", buf, pga_port(a));
	} else if (pga_family(a) == AF_UNIX && a->scred.pid) {
		snprintf(dst, dstlen, "%s(%lu@%s):%d", buf, (unsigned long)a->scred.pid, cached_hostname(), pga_port(a));
	} else {
		snprintf(dst, dstlen, "%s:%d", buf, pga_port(a));
	}
	return dst;
}

/* hints for getaddrinfo(listen_addr) */
static const struct addrinfo hints = {
	.ai_family = AF_UNSPEC,
	.ai_socktype = SOCK_STREAM,
	.ai_protocol = IPPROTO_TCP,
	.ai_flags = AI_PASSIVE,
};

/*
 * initialize another listening socket.
 */
static struct ListenSocket *init_socket(int af, const struct sockaddr *sa, int salen)
{
	struct ListenSocket *ls;
	int sock, res;
	char buf[128];
	const char *errpos;

	log_debug("init_socket: %s", sa2str(sa, buf, sizeof(buf)));

	/* create socket */
	errpos = "socket";
	sock = socket(af, SOCK_STREAM, 0);
	if (sock < 0)
		goto failed;

	/* SO_REUSEADDR behaviour it default in WIN32.  */
#ifndef WIN32
	/* relaxed binding */
	if (af != AF_UNIX) {
		int val = 1;
		errpos = "setsockopt";
		res = setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val));
		if (res < 0)
			goto failed;
	}
#endif

#ifdef IPV6_V6ONLY
	/* avoid ipv6 socket's attempt to takeover ipv4 port */
	if (af == AF_INET6) {
		int val = 1;
		errpos = "setsockopt/IPV6_V6ONLY";
		res = setsockopt(sock, IPPROTO_IPV6, IPV6_V6ONLY, &val, sizeof(val));
		if (res < 0)
			goto failed;
	}
#endif

	/* bind it */
	errpos = "bind";
	res = bind(sock, sa, salen);
	if (res < 0)
		goto failed;

	/* set common options */
	errpos = "tune_socket";
	if (!tune_socket(sock, (af == AF_UNIX)))
		goto failed;

	/* finally, accept connections */
	errpos = "listen";
	res = listen(sock, cf_listen_backlog);
	if (res < 0)
		goto failed;

	if (af == AF_UNIX) {
		struct sockaddr_un *un = (struct sockaddr_un *)sa;
		change_file_mode(un->sun_path, cf_unix_socket_mode, NULL, cf_unix_socket_group);
	} else {
		tune_accept(sock, cf_tcp_defer_accept);
	}

	errpos = "calloc";
	ls = calloc(1, sizeof(*ls));
	if (!ls)
		goto failed;

	list_init(&ls->node);
	ls->fd = sock;
	if (sa->sa_family == AF_UNIX) {
		pga_set(&ls->addr, AF_UNIX, cf_listen_port);
	} else {
		pga_copy(&ls->addr, sa);
	}

	if (af == AF_UNIX) {
		struct sockaddr_un *un = (struct sockaddr_un *)sa;
		change_file_mode(un->sun_path, cf_unix_socket_mode, NULL, cf_unix_socket_group);
	} else {
		tune_accept(sock, cf_tcp_defer_accept);
	}

	log_info("listening on %s", sa2str(sa, buf, sizeof(buf)));
	return ls;

failed:
	log_warning("Cannot listen on %s: %s(): %s",
		    sa2str(sa, buf, sizeof(buf)),
		    errpos, strerror(errno));
	if (sock >= 0)
		safe_close(sock);
	return NULL;
}

static bool parse_addr(void *arg, const char *addr)
{
	int res;
	char service[64];
	struct addrinfo *ai, *gaires = NULL;
	struct ListenSocket *ls;
	struct StatList *sock_list = arg;

	if (!*addr)
		return true;
	if (strcmp(addr, "*") == 0)
		addr = NULL;
	snprintf(service, sizeof(service), "%d", cf_listen_port);

	res = getaddrinfo(addr, service, &hints, &gaires);
	if (res != 0) {
		fatal("getaddrinfo('%s', '%d') = %s [%d]", addr ? addr : "*",
		      cf_listen_port, gai_strerror(res), res);
	}

	for (ai = gaires; ai; ai = ai->ai_next) {
		ls = init_socket(ai->ai_family, ai->ai_addr, ai->ai_addrlen);
		if (ls)
			statlist_append(sock_list, &ls->node);
		/* it's unclear whether all or only first result should be used */
	}

	freeaddrinfo(gaires);
	return true;
}

static struct ListenSocket *create_unix_socket(void)
{
	struct sockaddr_un un;
	int res;
	char lockfile[sizeof(struct sockaddr_un) + 10];
	struct stat st;

	/* fill sockaddr struct */
	memset(&un, 0, sizeof(un));
	un.sun_family = AF_UNIX;
	snprintf(un.sun_path, sizeof(un.sun_path),
		"%s/.s.PGSQL.%d", cf_unix_socket_dir, cf_listen_port);

	/* check for lockfile */
	snprintf(lockfile, sizeof(lockfile), "%s.lock", un.sun_path);
	res = lstat(lockfile, &st);
	if (res == 0)
		fatal("unix port %d is in use", cf_listen_port);

	/* expect old bouncer gone */
	unlink(un.sun_path);

	return init_socket(AF_UNIX, (const struct sockaddr *)&un, sizeof(un));
}

void create_sockets(struct StatList *sock_list)
{
	bool ok;
	struct ListenSocket *ls;

	if (cf_listen_addr && *cf_listen_addr) {
		ok = parse_word_list(cf_listen_addr, parse_addr, sock_list);
		if (!ok)
			fatal("failed to parse listen_addr list: %s", cf_listen_addr);
	}

	if (cf_unix_socket_dir && *cf_unix_socket_dir) {
		ls = create_unix_socket();
		if (ls)
			statlist_append(sock_list, &ls->node);
	}
}

void cleanup_sockets(struct StatList *sock_list)
{
	struct ListenSocket *ls;
	struct List *el;

	while ((el = statlist_pop(sock_list)) != NULL) {
		ls = container_of(el, struct ListenSocket, node);
		if (ls->fd > 0) {
			safe_close(ls->fd);
			ls->fd = 0;
		}
		if (pga_is_unix(&ls->addr)) {
			char buf[sizeof(struct sockaddr_un) + 20];
			snprintf(buf, sizeof(buf), "%s/.s.PGSQL.%d", cf_unix_socket_dir, cf_listen_port);
			unlink(buf);
		}
		statlist_remove(sock_list, &ls->node);
		free(ls);
	}
}

/*
 * Notify pooler only when also data is arrived.
 *
 * optval specifies how long after connection attempt to wait for data.
 *
 * Related to tcp_synack_retries sysctl, default 5 (corresponds 180 secs).
 *
 * SO_ACCEPTFILTER needs to be set after listen(), maybe TCP_DEFER_ACCEPT too.
 */
void tune_accept(int sock, bool on)
{
	const char *act = on ? "install" : "uninstall";
	int res = 0;
#ifdef TCP_DEFER_ACCEPT
	int val = 45; /* FIXME: proper value */
	socklen_t vlen = sizeof(val);
	res = getsockopt(sock, IPPROTO_TCP, TCP_DEFER_ACCEPT, &val, &vlen);
	log_noise("old TCP_DEFER_ACCEPT on %d = %d", sock, val);
	val = on ? 1 : 0;
	log_noise("%s TCP_DEFER_ACCEPT on %d", act, sock);
	res = setsockopt(sock, IPPROTO_TCP, TCP_DEFER_ACCEPT, &val, sizeof(val));
#else
#if 0
#ifdef SO_ACCEPTFILTER
	struct accept_filter_arg af, *afp = on ? &af : NULL;
	socklen_t af_len = on ? sizeof(af) : 0;
	memset(&af, 0, sizeof(af));
	strcpy(af.af_name, "dataready");
	log_noise("%s SO_ACCEPTFILTER on %d", act, sock);
	res = setsockopt(sock, SOL_SOCKET, SO_ACCEPTFILTER, afp, af_len);
#endif
#endif
#endif
	if (res < 0)
		log_warning("tune_accept: %s TCP_DEFER_ACCEPT/SO_ACCEPTFILTER: %s",
			    act, strerror(errno));
}
